msgraphcore/graph_session.py

```diff
@@ -20,47 +25,84 @@ def __init__(self, auth_provider: AuthProviderBase, middleware: list = []):
 
         auth_handler = AuthorizationHandler(auth_provider)
 
+        # The authorization handler should be the first middleware in the pipeline.
         middleware.insert(0, auth_handler)
         self._register(middleware)
 
-    def get(self, url: str, **kwargs) -> Response:
-        return self._prepare_and_send_request('GET', url, **kwargs)
-
-    def post(self, url: str, **kwargs) -> Response:
-        return self._prepare_and_send_request('POST', url, **kwargs)
-
-    def put(self, url: str, **kwargs) -> Response:
-        return self._prepare_and_send_request('PUT', url, **kwargs)
-
-    def patch(self, url: str, **kwargs) -> Response:
-        return self._prepare_and_send_request('PATCH', url, **kwargs)
-
-    def delete(self, url: str, **kwargs) -> Response:
-        return self._prepare_and_send_request('DELETE', url, **kwargs)
-
-    def _get_url(self, url: str) -> Response:
+    @middleware_control.get_middleware_options
```

Comment:
>@middleware_control.get_middleware_options [](start = 4, length = 42)

since we are stripping off all the options before calling get, etc.  it means we must know all members of the dictionary in order to populate options.  This also means that a user cannot create a custom middleware which needs to pass arguments and receive them unless the middlewarecontrol is extensible.

Reply:
That is correct, we know all options that middleware expect. If a user creates a custom middleware they can make it allow options through a constructor. 

Note: Other SDKs don't support user defined middleware

Reply:
it's a thing we should discuss but not blocking this PR.